src/state.rs

```diff
@@ -0,0 +1,219 @@
+            our_part: opt_part.ok_or(Error::NotInPubKeySet)?,
+            rng: rng.clone(),
+        })
+    }
+
+    /// The 1st vote with our Part
+    pub fn first_vote(&mut self) -> Result<DkgSignedVote> {
+        let vote = DkgVote::SinglePart(self.our_part.clone());
+        self.cast_vote(vote)
+    }
+
+    fn get_validated_vote(&self, vote: &DkgSignedVote) -> Result<DkgVote> {
+        let sender_id = vote.voter;
+        let sender_pub_key = self.pub_keys.get(&sender_id).ok_or(Error::UnknownSender)?;
+        let vote = vote.get_validated_vote(sender_pub_key)?;
+        Ok(vote)
+    }
+
+    fn all_checked_votes(&self) -> Result<Vec<(DkgVote, NodeId)>> {
+        self.all_votes
+            .iter()
+            .map(|v| Ok((self.get_validated_vote(v)?, v.voter)))
+            .collect()
+    }
+
+    fn current_dkg_state(&self, votes: Vec<(DkgVote, NodeId)>) -> DkgCurrentState {
+        let knowledge = match Knowledge::from_votes(votes) {
+            Err(KnowledgeFault::IncompatibleAcks) | Err(KnowledgeFault::IncompatibleParts) => {
+                return DkgCurrentState::IncompatibleVotes;
+            }
+            Err(KnowledgeFault::MissingParts) | Err(KnowledgeFault::MissingAcks) => {
+                return DkgCurrentState::NeedAntiEntropy;
+            }
+            Ok(k) => k,
+        };
+
+        let participants_len = self.pub_keys.len();
+        if knowledge.agreed_with_all_acks.len() == participants_len {
+            DkgCurrentState::Termination(knowledge.part_acks)
+        } else if !knowledge.agreed_with_all_acks.is_empty() {
+            DkgCurrentState::WaitingForTotalAgreement
+        } else if knowledge.got_all_acks(participants_len) {
+            DkgCurrentState::GotAllAcks(knowledge.part_acks)
+        } else if !knowledge.part_acks.is_empty() {
+            DkgCurrentState::WaitingForMoreAcks
+        } else if knowledge.parts.len() == participants_len {
+            DkgCurrentState::GotAllParts(knowledge.parts)
+        } else {
+            DkgCurrentState::WaitingForMoreParts
+        }
+    }
+
+    /// Sign, log and return the vote
+    fn cast_vote(&mut self, vote: DkgVote) -> Result<DkgSignedVote> {
+        let sig = self.secret_key.sign(&bincode::serialize(&vote)?);
+        let signed_vote = DkgSignedVote::new(vote, self.id, sig);
+        self.all_votes.insert(signed_vote.clone());
+        Ok(signed_vote)
+    }
+
+    /// Handles all the Acks
+    fn handle_all_acks(&mut self, all_acks: BTreeMap<IdPart, BTreeSet<IdAck>>) -> Result<()> {
+        for ((part_id, _part), acks) in all_acks {
+            for (sender_id, ack) in acks {
+                let outcome = self.keygen.handle_ack(&sender_id, ack.clone())?;
+                if let AckOutcome::Invalid(fault) = outcome {
+                    return Err(Error::FaultyVote(format!(
+                        "Ack fault: {:?} by {:?} for part by {:?}",
+                        fault, sender_id, part_id
+                    )));
```

Comment:
You can reduce these 4 lines down to 1 if you put the variables into the format string i.e. `"... {fault:?} ..."`